atomic_reactor/plugin.py

```diff
@@ -354,6 +391,44 @@ def __init__(self, dt, workflow, plugins_conf, *args, **kwargs):
         self.plugins_results = workflow.prebuild_results
         super(PreBuildPluginsRunner, self).__init__(dt, workflow, 'PreBuildPlugin', plugins_conf, *args, **kwargs)
 
+
+class BuildStepPlugin(BuildPlugin):
+    pass
+
+
+class BuildStepPluginsRunner(BuildPluginsRunner):
+
+    def __init__(self, dt, workflow, plugin_conf, *args, **kwargs):
+        logger.info("initializing runner of build-step plugin")
+        self.plugins_results = workflow.buildstep_result
+
+        if not plugin_conf:
+            # if none buildstep plugins specified, fallback to docker api plugin
+            plugin_conf = [{'name': 'docker_api', 'is_allowed_to_fail': False}]
+        else:
+            # any non existing buildstep plugin must be skipped without error
+            for i in range(len(plugin_conf)):
+                plugin_conf[i]['required'] = False
+                plugin_conf[i]['is_allowed_to_fail'] = False
+
+        super(BuildStepPluginsRunner, self).__init__(
+            dt, workflow, 'BuildStepPlugin', plugin_conf, *args, **kwargs)
+
+    def run(self, *args, **kwargs):
+        builder = self.workflow.builder
+
+        logger.info('building image %r inside current environment',
+                    builder.image)
+        builder._ensure_not_built()
+        logger.debug('using dockerfile:\n%s',
+                     DockerfileParser(builder.df_path).content)
+
+        kwargs['buildstep_phase'] = True
+
+        plugins_results = super(BuildStepPluginsRunner, self).run(*args, **kwargs)
+        return plugins_results
```

Comment:
Since this class only ever returns a single result, how about just returning that single result?
```
return plugins_results.values()[0]
```

This would allow build_docker_image() to continue to return a BuildResult instead of changing it to return a list (and have to modify callers to deal with the list).